prov/efa/src/efa_base_ep.c

```diff
@@ -186,12 +186,29 @@ int efa_qp_create(struct efa_qp **qp, struct ibv_qp_init_attr_ex *init_attr_ex,
 #endif
 		efa_attr.driver_qp_type = EFADV_QP_DRIVER_TYPE_SRD;
 #if HAVE_EFADV_SL
-		if (tclass == FI_TC_LOW_LATENCY)
-			efa_attr.sl = EFA_QP_DEFAULT_SERVICE_LEVEL;
-#endif
+		efa_attr.sl = EFA_QP_DEFAULT_SERVICE_LEVEL;
+		if (tclass == FI_TC_LOW_LATENCY) {
+			efa_attr.sl = EFA_QP_LOW_LATENCY_SERVICE_LEVEL;
+			(*qp)->ibv_qp = efadv_create_qp_ex(
+				init_attr_ex->pd->context, init_attr_ex, &efa_attr,
+				sizeof(struct efadv_qp_init_attr));
+			if (!(*qp)->ibv_qp) {
+				EFA_INFO(FI_LOG_EP_CTRL, "ibv_create_qp failed with sl %u, errno: %d. Retrying with default sl. \n", efa_attr.sl, errno);
+				efa_attr.sl = EFA_QP_DEFAULT_SERVICE_LEVEL;
+				(*qp)->ibv_qp = efadv_create_qp_ex(
+					init_attr_ex->pd->context, init_attr_ex, &efa_attr,
+					sizeof(struct efadv_qp_init_attr));
+			}
+		} else {
+			(*qp)->ibv_qp = efadv_create_qp_ex(
+				init_attr_ex->pd->context, init_attr_ex, &efa_attr,
+				sizeof(struct efadv_qp_init_attr));
+		}
+#else
 		(*qp)->ibv_qp = efadv_create_qp_ex(
 			init_attr_ex->pd->context, init_attr_ex, &efa_attr,
 			sizeof(struct efadv_qp_init_attr));
```

Comment:
Can we refactor this s.t. we're not repeating the same thing 4 times here (`(*qp)->ibv_qp = efadv_create_qp_ex(...)`)?

Reply:
I thought about it, but the current version is the what I think the most intuitive one. Any suggestions?

Reply:
We may move the retry logic to https://github.com/ofiwg/libfabric/blob/main/prov/efa/src/efa_base_ep.c#L197. It can reduce the number of `efadv_create_qp_ex` calls

Reply:
But it will involve more `HAVE_EFADV_SL` macros ... which I don''t think makes the code cleaner.

Reply:
Like
```
#if HAVE_EFADV_SL
	efa_attr.sl = (tclass == FI_TC_LOW_LATENCY) 
		? EFA_QP_LOW_LATENCY_SERVICE_LEVEL 
		: EFA_QP_DEFAULT_SERVICE_LEVEL;
#endif
(*qp)->ibv_qp = efadv_create_qp_ex()

if (!(*qp)->ibv_qp) {
#if HAVE_EFADV_SL
        if (tclass == FI_TC_LOW_LATENCY) {
                // retry
        }
#endif
...
}
```

Reply:
Updated